Pycuampcor v2

lijun99 opened the pull request:
PyCuAmpcor Version 2 introduces a new workflow inspired by [GrIMP/SpeckleSource](https://github.com/fastice/speckleSource), in addition to the original ROIPAC/ampcor workflow. Below is an overview of the new workflow and additional improvements:

Key Features of the New GrIMP/SpeckleSource Workflow:
 - Starts with an anti-aliasing oversampled secondary chip over the entire search range.
 - Adds extra margins to facilitate correlation surface extraction.
 - Offers optional double-precision computations (also extended to the ROIPAC/ampcor workflow for consistency).

The GrIMP/SpeckleSource workflow may be slower than ROIPAC/ampcor due to increased computational demands.
However, it may improve accuracy in scenarios with low signal-to-noise ratio (SNR). More details on the workflows and how to choose can be found at contrib/PyCuAmpcor/README.md. 

Other Improvements
 - Introduced an option to fix the starting pixel location, which is ideal for comparing results from different window sizes.
 - Added an example script for plotting the offset field, making it easier for users to visualize results.
 - Included correlation peak values in the output, alongside SNR and covariance, providing additional accuracy measures.
 - Resolved a memory deallocation issue that occurred when the run procedure was called repeatedly within a single script.

Commit:
PyCuAmpcor: offer an option to fix the starting pixel locations

in this update,
* either the starting pixel(corner) or the center of window location may be specified (see README)
* by specifying the starting/ending pixel locations, the reference/search window may go beyond the range of the original images. Zeros are filled in instead. In this case, the offset fields at the edge may be incorrect.
* The range check is no longer needed. It only runs in DEBUG mode and issues warning messages when pixels are outside the image range.

Commit:
PyCuAmpcor: merge #622 by rtburns-jpl

Commit:
pycuampcor: deallocate memory objects

Commit:
Merge branch 'main' into pycuampcor

Commit:
pycuampcor: add debug messages for loading the images

Commit:
PyCuAmpcor V2: start with oversampled images

Commit:
PyCuAmpcor: fix the formula for determining the final offset

Commit:
PyCuAmpcor: oversampling procedure adjustment

Commit:
PyCuAmpcor: add extra pads to correlation surface

Commit:
PyCuAmpcor: use split spectrum for even sequence oversampling

Commit:
PyCuAmpcor: add margin to the loaded secondary chip as well

Commit:
PyCuAmpcor: add a python script to plot offset fields

Commit:
PyCuAmpcor: use a larger area to oversample the reference image

Commit:
PyCuAmpcor: first attemp with double precision - use CUAMPCOR_DOUBLE flag

Commit:
PyCuAmpcor-DB: first running version

Commit:
PyCuAmpcor-DB: fixed some errors to produce the correct results

Commit:
PyCuAmpcor-DP: disable time domain algorithm for (os) window larger than 640 with DP

Commit:
PyCuAmpcor-DP: change default to single precision

Commit:
PyCuAmpcor-v2: solution to time domain code for gpus with 48K shared memory

Commit:
PyCuAmpcor-V2: in zero padding for FFT oversampling, resort to conventional approach to side Nyquist frequency (N/2) with negative frequency

Commit:
PyCuAmpcor-v2: add correlation surface statistics (snr and variance) back

Commit:
PyCuAmpcor-v2: add the correlation surface peak values to output

Commit:
PyCuAmpcor-v2: rename cuAmpcorChunk to cuAmpcorProcessor and use the virtual class for different workflows

Commit:
PyCuAmpcor-v2: bring back ROIPAC/ampcor workflow (set as default), bring back Sinc Oversampler

Commit:
PyCuAmpcor-v2: repair Readme

hfattahi commented:
Thank you @lijun99 for this extensive PR. Based on our offline discussions I'm aware of most of developments in this PR. However, for clarity and to help the user community following this repository would you please update the description of the PR and provide some explanation in few bullets summarizing the changes that comes in this PR.

vbrancat commented on contrib/PyCuAmpcor/src/cuEstimateStats.cu:
```diff
@@ -58,72 +90,73 @@ __global__ void cudaKernel_estimateVar(const float* corrBatchRaw, const int NX,
-        int idx21 = offset + (px + 1) * NY + py    ;
-        int idx22 = offset + (px + 1) * NY + py + 1;
+        int idx12 = offset + (px    ) * NY + py + distance;
+        int idx20 = offset + (px + distance) * NY + py - distance;
+        int idx21 = offset + (px + distance) * NY + py    ;
+        int idx22 = offset + (px + distance) * NY + py + distance;
 
         // second-order derivatives
-        float dxx = - ( corrBatchRaw[idx21] + corrBatchRaw[idx01] - 2.0*corrBatchRaw[idx11] );
-        float dyy = - ( corrBatchRaw[idx12] + corrBatchRaw[idx10] - 2.0*corrBatchRaw[idx11] ) ;
-        float dxy = ( corrBatchRaw[idx22] + corrBatchRaw[idx00] - corrBatchRaw[idx20] - corrBatchRaw[idx02] ) *0.25;
+        real_type dxx = - ( corrBatchRaw[idx21] + corrBatchRaw[idx01] - 2.0*corrBatchRaw[idx11] );
+        real_type dyy = - ( corrBatchRaw[idx12] + corrBatchRaw[idx10] - 2.0*corrBatchRaw[idx11] ) ;
+        real_type dxy = ( corrBatchRaw[idx22] + corrBatchRaw[idx00] - corrBatchRaw[idx20] - corrBatchRaw[idx02] ) *0.25;
 
-        float n2 = fmaxf(1.0 - peak, 0.0);
+        real_type n2 = max(1.0 - peak, 0.0);
 
         dxx = dxx * templateSize;
         dyy = dyy * templateSize;
         dxy = dxy * templateSize;
 
-        float n4 = n2*n2;
+        real_type n4 = n2*n2;
         n2 = n2 * 2;
         n4 = n4 * 0.5 * templateSize;
 
-        float u = dxy * dxy - dxx * dyy;
-        float u2 = u*u;
+        real_type u = dxy * dxy - dxx * dyy;
+        real_type u2 = u*u;
 
         // if the Gaussian curvature is too small
         if (fabsf(u) < 1e-2) {
-            covValue[idxImage] = make_float3(99.0, 99.0, 0.0);
+            covValue[idxImage] = make_real3(99.0, 99.0, 0.0);
```
To my understanding, if the Gaussian curvature is too small (less than 0.01) then we assign a `noData` value of 99. Why we do this and where is the threshold of 0.01 come from?

lijun99 replied:
This follows the ROIPAC. We could make it a user configurable parameter.

vbrancat reviewed (commented):
@lijun99 Thank you for this PR! It’s very exciting, and I’m eager to test it. I will leave some further comments after I am done with the testing.

 In the meantime, I’ve added some comments and questions on the code. Since I’m not a CUDA programmer, I’ll leave it to @gmgunter  and @rtburns-jpl  to weigh in on the CUDA-specific parts. Overall, the PR looks great to me, but there are a few comments and considerations that I think need to be addressed:

1. I’m not a fan of the ROIPAC vs. GRiMP terminology, as I believe most users won’t be familiar with it either. At a high level (e.g., in the Python interface), we should frame this in terms of enabling or disabling anti-aliasing oversampling and provide a clear explanation of the advantages and disadvantages of each option. This approach would ensure complete transparency for users.

2. It’s unclear to me how offset estimates are handled near the edges. For instance, if we’re at the edge of the search window, do we use zero-padding? Could you point me to the part of the code where this logic is implemented?

3. I noticed that the code internally includes a threshold on the SNR, but it’s unclear to me if this option is actively used. Additionally, this threshold is not exposed to users. Could you clarify its purpose and usage?

hfattahi commented:
@lijun99 Thank you for adding the PR description. 
I agree with @vbrancat 's points about the GRIMP/ROIPAC terminology. Perhaps the correct thing is to be just clear about  the changes made to the PyCuAmpcor. Main reason is that the base of GRIMP was also original JPL's roipac ampcor. The fact that upfront upsampling has been dropped at some point in some version of it for computational efficiency while kept in other versions such as the one in GRIMP may be just a source of confusion. 

We see that one big change contributing to the big PR is the functionality to build the code with both single/double precision. Since our tests have not demonstrated a meaningful difference for the offsets results with the two builds, we are a bit hesitant if we should go that way. Or at least we are not sure if the implementation approach for adding double precision is the most efficient way. Are you open to:

1- remove double precision build from this PR. We expect that would significantly reduce the number of lines of codes in this PR
2- If you definitely want to add double precision and have a good reason for it, let's discuss if templating the code would be a better approach and have a separate PR in future.

Commit:
PyAcuAmpcor-v2: add debug info to cuderamp=1

Commit:
PyAcuAmpcor-v2: fix the search range in presence of extra pads

Commit:
PyAcuAmpcor-v2: rename two workflows to TwoPass and OnePass wrt the number of NCC passes

Commit:
PyCuAmpcor-v2: clarify notations and notes, per PR comments

Commit:
PyCuAmpcor-v2: add description for the offset visualization tool plotOffsets.py

lijun99 commented on contrib/PyCuAmpcor/README.md:
```diff
@@ -83,25 +93,30 @@ ww=64  # template window width
 
 rm $outprefix$outsuffix*
-cuDenseOffsets.py --reference $reference --secondary $secondary --ww $ww --wh $wh --sw $sw --sh $sh --mm $mm --kw $kw --kh $kh --gross $gross --rr $rgshift --aa $azshift --oo $oo --deramp $deramp --outprefix $outprefix --outsuffix $outsuffix --gpuid $gpuid  --usemmap $usemmap --mmapsize $mmapsize --nwac $nwac --nwdc $nwdc 
+cuDenseOffsets.py --reference $reference --secondary $secondary --ww $ww --wh $wh --sw $sw --sh $sh \
+    --mm $mm --kw $kw --kh $kh --gross $gross --rr $rgshift --aa $azshift --oo $oo \
+    --deramp $deramp --outprefix $outprefix --outsuffix $outsuffix \
+    --gpuid $gpuid  --usemmap $usemmap --mmapsize $mmapsize \
+    --nwac $nwac --nwdc $nwdc --workflow $workflow
```
added.

lijun99 commented on contrib/PyCuAmpcor/examples/cuDenseOffsets.py:
```diff
@@ -79,14 +83,26 @@ def createParser():
+    parser.add_argument('--startpixeldw', dest='startpixeldw', type=int, default=None,
                         help='Starting Pixel down of the reference image.' +
-                             'Default: %(default)s to be determined by margin and search range.')
+                             'Default: %(default)s to use the first pixel.')
```
This is to fix the location of the reference window, for cases to compare results with different window sizes and search ranges. It could lead to out-of-range issues for windows at the edges. Zeros are padded to keep the program running. Warnings are raised to users to alert them of the inaccuracy of these results.

lijun99 commented on contrib/PyCuAmpcor/examples/cuDenseOffsets.py:
```diff
@@ -79,14 +83,26 @@ def createParser():
 
     # cross-correlation algorithm
+    parser.add_argument('--wf', '--workflow', dest='workflow', type=int, default=0,
+                        help='workflow (0 = ROIPAC, 1 = GrIMP) (default: %(default)s).')
```
added

lijun99 commented on contrib/PyCuAmpcor/examples/cuDenseOffsets.py:
```diff
@@ -234,20 +244,80 @@ def estimateOffsetField(reference, secondary, inps=None):
+    if inps.startpixeldw is not None:
+        # use the given starting pixel coordinate
+        startPixelDown = inps.startpixeldw
+        if inps.use_center != 0:
+            # the given starting coordinate is the center, adjust it
+            startPixelDown -= inps.winhgt//2
```
Explained earlier.

lijun99 commented on contrib/PyCuAmpcor/examples/cuDenseOffsets.py:
```diff
@@ -234,20 +244,80 @@ def estimateOffsetField(reference, secondary, inps=None):
+            # the given starting coordinate is the center, adjust it
+            startPixelDown -= inps.winhgt//2
+    else:
+        # use margin + halfSearchHeight instead
+        startPixelDown = margin + inps.srchgt
```
explained earlier.

lijun99 commented on contrib/PyCuAmpcor/examples/cuDenseOffsets.py:
```diff
@@ -234,20 +244,80 @@ def estimateOffsetField(reference, secondary, inps=None):
+    if endPixelAcross > width - inps.srcwidth:
+        print(f'Warning: the ending pixel across {endPixelAcross} is more than width {width} - search range {inps.srcwidth} ')
+
     # oversample raw data (SLC)
     objOffset.rawDataOversamplingFactor = inps.raw_oversample
```
Renamed for clarification.

lijun99 commented on contrib/PyCuAmpcor/examples/cuDenseOffsets.py:
```diff
@@ -268,10 +338,12 @@ def estimateOffsetField(reference, secondary, inps=None):
     objOffset.grossOffsetImageName = fbase + '_gross.bip'
     objOffset.snrImageName = fbase + '_snr.bip'
     objOffset.covImageName = fbase + '_cov.bip'
+    objOffset.peakValueImageName = fbase + '_peak.bip'
```
yes, it is normalized value.

lijun99 commented on contrib/PyCuAmpcor/examples/cuDenseOffsets.py:
```diff
@@ -342,7 +417,8 @@ def estimateOffsetField(reference, secondary, inps=None):
             return objOffset, geomDict
 
     # Run the code
-    print('Running PyCuAmpcor')
+    Precision = 'DOUBLE' if objOffset.isDoublePrecision() else 'FLOAT'
+    print(f'Running PyCuAmpcor in {Precision} precision')
```
Currently, the double precision option can only be turned on from compiling. Reserved for developers. We'd better not to expose it to common users for now.

lijun99 commented on contrib/PyCuAmpcor/src/cuAmpcorParameter.h:
```diff
@@ -71,9 +74,23 @@ class cuAmpcorParameter{
+    // correlation surface
+    int2 corrWindowSize; // 2*halfSearchRange + 1
+    int2 corrZoomInSize; // zoomWindowSize+1
+    int2 corrZoomInOversampledSize; // corrZoomInSize * oversamplingFactor
```
Renamed.

lijun99 commented on contrib/PyCuAmpcor/src/cuAmpcorParameter.h:
```diff
@@ -71,9 +74,23 @@ class cuAmpcorParameter{
+    int2 corrZoomInSize; // zoomWindowSize+1
+    int2 corrZoomInOversampledSize; // corrZoomInSize * oversamplingFactor
+
+    int corrStatWindowSize;     ///< correlation surface size used to estimate snr
```
covariance only needs a few pixels around the peak. The statement was correct.

lijun99 commented on contrib/PyCuAmpcor/src/cuAmpcorParameter.h:
```diff
@@ -71,9 +74,23 @@ class cuAmpcorParameter{
+
+    int corrStatWindowSize;     ///< correlation surface size used to estimate snr
 
     float thresholdSNR;      ///< Threshold of Signal noise ratio to remove noisy data
```
Deleted. PyCuAmpcor leaves filtering, cullings for users.

lijun99 commented on contrib/PyCuAmpcor/src/cuAmpcorParameter.h:
```diff
@@ -125,10 +142,14 @@ class cuAmpcorParameter{
     std::string offsetImageName;       ///< Offset fields output filename
     std::string snrImageName;          ///< Output SNR filename
     std::string covImageName;          ///< Output variance filename
+    std::string peakValueImageName;      ///< Output correlation surface peak value filename
```
Changed.

lijun99 commented on contrib/PyCuAmpcor/examples/cuDenseOffsets.py:
```diff
@@ -23,11 +23,11 @@
   # offset and its geometry
   # tip: re-run with --full/out-geom and without --redo to generate geometry only
-  cuDenseOffsets.py -r ./SLC/20151120/20151120.slc.full -s ./SLC/20151214/20151214.slc.full --outprefix ./offsets/20151120_20151214/offset --ww 256 --wh 256 --sw 8 --sh 8 --oo 32 --kw 300 --kh 100 --nwac 100 --nwdc 1 --gpuid 2 --full-geom ./geom_reference --out-geom ./offset/geom_reference
+  cuDenseOffsets.py -r ./SLC/20151120/20151120.slc.full -s ./SLC/20151214/20151214.slc.full --outprefix ./offsets/20151120_20151214/offset --ww 256 --wh 256 --sw 8 --sh 8 --oo 32 --kw 300 --kh 100 --nwac 100 --nwdc 1 --gpuid 0 --full-geom ./geom_reference --out-geom ./offset/geom_reference
```
Done.

Commit:
PyCuAmpcor-v2: revert to custom mmap image loader

Commit:
PyCuAmpcor-v2: 1) fix data types for DP; 2) reduce the search range of the oversampled correlation surface

Commit:
pycuampcor-v2: fix an error in determining the corrwindow size

Commit:
pycuampcor-v2: fix also the corrwindow size error for two-pass workflow

Commit:
pycuampcor-v2: add get_sm_count method - the batch size is recommended to be a factor of #SMs

Commit:
patch getLineBand in some fortran code

Commit:
Merge branch 'main' into pycuampcor-v2-devel

Commit:
Merge branch 'fortran-patch' into pycuampcor-v2-devel

Commit:
PyCuAmpcor: patch the scons script for the two different vironments issue

Commit:
pycuampcor-v2: change the order of parameters in classes to suppress the warnings generated by -Wreorder

yuankailiu committed:
pycuAmpcor add new workflow to TopsProc, topsApp

+ self.workflow either 0 (old two-step) or 1 (new one-step)
+ update this option for both cpu and gpu function
+ add workflow keyword into applications.topsApp.TopsInSAR

Commit:
pycuampcor for topsApp: change the slc access method; add the output for correlation peak values

yuankailiu committed:
add OFFSET_PEAKVALUE_FILE to TopsProc

Commit:
pycuampcor in topsApp: fix the runDenseOffsets script

yuankailiu committed:
revert typo changes

Commit:
pycuampcor-v2: fix the distance in variance estimate in one-pass workflow

lijun99 commented:
I have merged all the pycuampcor updates to [a separate repo](https://github.com/lijun99/cuAmpcor), which can be embedded as a submodule in isce2/isce3: here is [an example](https://github.com/lijun99/isce2/commit/975021568c35c5c9cdfdd8b73a0b970e78ea6acc). @rtburns-jpl , please let me know whether this is what you prefer. ( I have renamed the module in lower cases, per the conda requirement).
